core/tee/tee_sql_fs.c

```diff
@@ -0,0 +1,879 @@
+{
+	struct sql_fs_fd *fdp = handle_lookup(&fs_db, fd);
+	tee_fs_off_t ret = -1;
+	tee_fs_off_t raw_pos;
+	tee_fs_off_t pos;
+
+	DMSG("(fd: %d, offset: %" PRId64 ", whence: %d)...", fd, offset,
+	     whence);
+
+	if (!fdp)
+		goto exit_ret;
+
+	*errno = TEE_ERROR_GENERIC;
+
+	sql_fs_begin_transaction_rpc();
+
+	switch (whence) {
+	case TEE_FS_SEEK_SET:
+		pos = offset;
+		break;
+
+	case TEE_FS_SEEK_CUR:
+		pos = fdp->pos + offset;
+		break;
+
+	case TEE_FS_SEEK_END:
+		pos = fdp->meta.length + offset;
+		break;
+
+	default:
+		*errno = TEE_ERROR_BAD_PARAMETERS;
+		goto exit;
+	}
+
+	raw_pos = pos_to_raw(pos);
+	if (raw_pos < 0)
+		goto exit;
+	raw_pos = sql_fs_lseek_rpc(fd, raw_pos, TEE_FS_SEEK_SET);
+	if (raw_pos < 0)
+		goto exit;
+	ret = raw_to_pos(raw_pos);
+	if (ret < 0)
+		goto exit;
+	*errno = TEE_SUCCESS;
+exit:
+	sql_fs_end_transaction_rpc(ret < 0);
+	fdp->pos = ret;
+exit_ret:
+	DMSG("...%" PRId64, ret);
+	return ret;
+}
+
+static int sql_fs_close(int fd)
+{
+	struct sql_fs_fd *fdp = handle_lookup(&fs_db, fd);
+	int ret = -1;
+
+	DMSG("(fd: %d)...", fd);
+
+	if (!fdp)
+		goto exit;
+
+	sql_fs_close_rpc(fd);
+	put_fdp(fdp);
+
+	ret = 0;
+exit:
+	DMSG("...%d", ret);
+	return ret;
+}
+
+static int sql_fs_open(TEE_Result *errno, const char *file, int flags, ...)
+{
+	struct sql_fs_fd *fdp = NULL;
+	int rflags = flags | TEE_FS_O_RDWR; /* Need to read/write meta */
+	int exists;
+	int fd = -1;
+
+	DMSG("(file: %s, flags: %d)...", file, flags);
+
+	exists = !sql_fs_access_rpc(file, TEE_FS_F_OK);
+	if (flags & TEE_FS_O_CREATE) {
+		if ((flags & TEE_FS_O_EXCL) && exists) {
+			*errno = TEE_ERROR_ACCESS_CONFLICT;
+			goto exit;
+		}
+	} else {
+		if (!exists) {
+			*errno = TEE_ERROR_ITEM_NOT_FOUND;
+			goto exit;
+		}
+	}
+
+	fdp = (struct sql_fs_fd *)calloc(1, sizeof(*fdp));
+	if (!fdp) {
+		*errno = TEE_ERROR_OUT_OF_MEMORY;
+		goto exit;
+	}
+
+	fdp->fd = sql_fs_open_rpc(errno, file, rflags);
+	if (fdp->fd < 0)
+		goto exit;
+
+	fdp->flags = flags;
+
+	fd = read_meta(errno, fdp);
+	if (fd < 0)
+		goto exit;
+
+	fd = handle_get(&fs_db, fdp);
+
+exit:
+	if (fd < 0)
+		free(fdp);
+	DMSG("...%d", fd);
+	return fd;
+}
+
+static int sql_fs_read(TEE_Result *errno, int fd, void *buf, size_t len)
+{
+	struct sql_fs_fd *fdp = handle_lookup(&fs_db, fd);
+	size_t remain_bytes = len;
+	uint8_t *data_ptr = buf;
+	uint8_t *block = NULL;
+	int start_block_num;
+	int end_block_num;
+	int res = -1;
+	int ret;
+
+	DMSG("(fd: %d, buf: %p, len: %zu)...", fd, (void *)buf, len);
+
+	if (!fdp) {
+		*errno = TEE_ERROR_BAD_PARAMETERS;
+		goto exit;
+	}
+
+	if (fdp->pos + len > fdp->meta.length)
+		len = fdp->meta.length - fdp->pos;
+
+	if (!len) {
+		res = 0;
+		goto exit;
+	}
+
+	if (!buf) {
+		*errno = TEE_ERROR_BAD_PARAMETERS;
+		goto exit;
+	}
+
+	if (fdp->flags & TEE_FS_O_WRONLY) {
+		*errno = TEE_ERROR_ACCESS_CONFLICT;
+		goto exit;
+	}
+
+	start_block_num = block_num(fdp->pos);
+	end_block_num = block_num(fdp->pos + len + 1);
```

Comment:
if `(fdp->pos + len)` equals the start address of a block, `end_block_num = block_num(fdp->pos + len + 1);` will go 1 block ahead than expected.